exporter/elasticsearchexporter/internal/datapoints/histogram.go

```diff
@@ -0,0 +1,101 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package datapoints // import "github.com/open-telemetry/opentelemetry-collector-contrib/exporter/elasticsearchexporter/internal/datapoints"
+
+import (
+	"errors"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/exporter/elasticsearchexporter/internal/elasticsearch"
+)
+
+type Histogram struct {
+	pmetric.HistogramDataPoint
+	elasticsearch.MappingHintGetter
+	metric pmetric.Metric
+}
+
+func NewHistogram(metric pmetric.Metric, dp pmetric.HistogramDataPoint) Histogram {
+	return Histogram{
+		HistogramDataPoint: dp,
+		MappingHintGetter:  elasticsearch.NewMappingHintGetter(dp.Attributes()),
+		metric:             metric,
+	}
+}
+
+func (dp Histogram) Value() (pcommon.Value, error) {
+	if dp.HasMappingHint(elasticsearch.HintAggregateMetricDouble) {
+		vm := pcommon.NewValueMap()
+		m := vm.Map()
+		m.PutDouble("sum", dp.Sum())
+		m.PutInt("value_count", safeUint64ToInt64(dp.Count()))
+		return vm, nil
+	}
+	return histogramToValue(dp.HistogramDataPoint)
+}
+
+func (dp Histogram) DynamicTemplate(_ pmetric.Metric) string {
+	if dp.HasMappingHint(elasticsearch.HintAggregateMetricDouble) {
+		return "summary"
+	}
+	return "histogram"
+}
+
+func (dp Histogram) DocCount() uint64 {
+	return dp.HistogramDataPoint.Count()
+}
+
+func (dp Histogram) Metric() pmetric.Metric {
+	return dp.metric
+}
+
+func histogramToValue(dp pmetric.HistogramDataPoint) (pcommon.Value, error) {
+	// Histogram conversion function is from
+	// https://github.com/elastic/apm-data/blob/3b28495c3cbdc0902983134276eb114231730249/input/otlp/metrics.go#L277
```

Comment:
Is there no alternative to duplicating the logic here? Is there any way to make the original functionality public? Would it be helpful to at least add a link in the `histogramSample` method that's being duplicated to update this method whenever it changes?

I see this method is simply being copied, no need to block this PR, just thought I'd call it out.

Reply:
Possibly. I'm improving the overall readability and separation of the module, and it's already causing frequent conflicts with other changes. So'id rather avoid mixing refactorings into it.

Reply:
Totally fair, thanks!